### PR TITLE
TSCH stats and adaptive TSCH channel selection

### PR DESCRIPTION
**Pull request by atiselsts**
This PR adds TSCH statistics (for example, max synchronization error, number of disassociations, background noise levels per channel) and statistics for the TSCH timesource neighbor (per-channel ACK rate, per-channel packet RSSI and LQI).

It also adds adaptive channel selection service and examples that show how to enable the TSCH stats and the adaptive channel selection.

### CODE REVIEW

File: os/net/mac/tsch/tsch-stats.c [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+    TSCH_STATS_EWMA_UPDATE(stats->channel_stats[index].lqi,
+        TSCH_STATS_TRANSFORM(lqi, TSCH_STATS_LQI_SCALING_FACTOR));
+  }
+}
+/*---------------------------------------------------------------------------*/
+void
+tsch_stats_on_time_synchronization(int32_t sync_error)
+{
+  /* Update the maximal error so far if the absolute value of the new one is larger */
+  tsch_stats.max_sync_error = MAX(tsch_stats.max_sync_error, ABS(sync_error));
```
[the remark this answers is not in the document]
Comment (atiselsts) [reply os/net/mac/tsch/tsch-stats.c]: No, this is sync error - yours is about drift.

Actually, would be good to unify this and other "statistics on the current session" variables? I mean, put them in TSCH stats. I would have done that originally I guess, but only code is older than these stats.

File: examples/6tisch/channel-selection-demo/node.c [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+  PROCESS_BEGIN();
+
+  is_coordinator = 0;
+
+#if CONTIKI_TARGET_COOJA
+  is_coordinator = (node_id == 1);
+#endif
+
+  /* Initialize the channel selection module */
+  cs_adaptations_init();
```
[the remark this answers is not in the document]
Comment (atiselsts) [reply examples/6tisch/channel-selection-demo/node.c]: Done

File: os/services/channel-selection/tsch-channel-selection.h [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+/* < 85% free */
+#define TSCH_CHANNEL_SELECTION_FREE_THRESHOLD ((tsch_stat_t)(85ul * TSCH_STATS_BINARY_SCALING_FACTOR / 100))
+#endif
+
+#define TSCH_CHANNEL_SELECTION_LEARNING_PERIOD_SEC 30
+
+/**
+ * \brief Initializes the TSCH hopping sequence selection module.
+ */
+void cs_adaptations_init(void);
```
[the remark this answers is not in the document]
Comment (atiselsts) [reply os/services/channel-selection/tsch-channel-selection.h]: Ok, done

File: os/net/mac/tsch/tsch-stats.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (simonduq) [reply os/net/mac/tsch/tsch-stats.c]: Unifying is good. I think "mine" actually also was about sync error, but was misnamed (just as the `drift_correction` variable in `tsch_timesync_update` and all sync error variables in `tsch-slot-operation.c` :s

[the remark this answers is not in the document]
Comment (simonduq) [reply os/net/mac/tsch/tsch-stats.c]: Let me know if you want to unify as part of this PR or would rather move it on in the current state. Both fine with me

[the remark this answers is not in the document]
Comment (atiselsts) [reply os/net/mac/tsch/tsch-stats.c]: I can do more changes later in a separate PR.

[the remark this answers is not in the document]
Comment (simonduq) [reply os/net/mac/tsch/tsch-stats.c]: ok

**Pull request merged**